pkg/asset/installconfig/aws/subnet.go

```diff
@@ -0,0 +1,165 @@
+package aws
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/ec2"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+)
+
+var vpc string
+var vpcFromSubnet string
+var routeTables []*ec2.RouteTable
+var subnetMetadata map[string]*Subnet
+
+// Subnet hold metadata for a subnet.
+type Subnet struct {
+	// ARN is the subnet's Amazon Resource Name.
+	ARN string
+
+	// Public is true if the cloud-provider logic considers the subnet
+	// to be public (e.g. with suitable routing for hosting public
+	// load balancers).
+	Public bool
+
+	// VPC is the ID of the VPC to which the subnet belongs.
+	VPC string
+
+	// Zone is the subnet's availability zone.
+	Zone string
+}
+
+// SubnetMetadata retrieves metadata for the given subnet(s).  Results
+// are cached, so additional requests for a given subnet do not result
+// in further AWS API calls.
+func SubnetMetadata(ctx context.Context, ids []string) (map[string]Subnet, error) {
```

Comment:
this function is complex enough that I think we should add unit test for this.

Also I would expect the client to be passed to this function rather than calling GetSession in this function.

Reply:
> this function is complex enough that I think we should add unit test for this.

What sort of framework do you see?  [The `EC2` client][1] is a `struct`, not an interface, so mocking seems hard.  Should I define my own `interface{}` so I can plug in a mock?  Should I recycle an existing mock?  Looks like there may be something [down deep in our vendored code][1], but seems pretty heavy to vendor that again to use in a unit test.  Personally, I think CI coverage for the bring-your-own-subnets flow will give sufficient confidence in how this is functioning.

> Also I would expect the client to be passed...

Do you expect our callers to call `GetSession`?  Or do I need to put a session on an asset so I only call `GetSession` once during asset generation?

[1]: https://github.com/openshift/installer/blob/f0ad4e618b5d46bfb2ea08e871165a5d30bf69c4/pkg/terraform/exec/plugins/vendor/github.com/hashicorp/aws-sdk-go-base/mock.go#L18

[1]: https://docs.aws.amazon.com/sdk-for-go/api/service/ec2/#EC2